### PR TITLE
Auto-inference of destination.service.resource

### PR DESCRIPTION
**Pull request by eyalkoren**
## What does this PR do?

Closing #1848 

## Checklist

- [x] This is an enhancement of existing features, or a new feature in existing plugins
  - [x] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-agent-java/blob/master/CHANGELOG.asciidoc)
  - [x] I have added tests that prove my fix is effective or that my feature works
  - [x] Added an API method or config option? Document in which version this will be introduced
  - [x] I have made corresponding changes to the documentation
  - [x] I added the test JSON spec to the APM repo

### CODE REVIEW

**Comment by eyalkoren** [comment]
I started also implementing the addition of the creation of exit spans, as described in the spec, but I think it doesn't make a lot of sense in the Java agent, as we don't expose any API to add custom data. A new API for exit span creation will look something like `startExitSpan(String type, String subtype, String action)`, which means that the `type` or `subtype` will be used for auto-inference of the `service.resource`, however this is essentially a capability we already offer through the API to set the service resource explicitly. @felixbarny WDYT?

**Comment by felixbarny** [comment]
> however this is essentially a capability we already offer through the API to set the service resource explicitly

The difference is that users don't need to remember setting resource fields or worry about which value to set in order for the span to show up in the service map.

File: apm-agent-core/src/test/resources/json-specs/service_resource_inference.json [hunk 1]
```diff
@@ -0,0 +1,183 @@
+[
```
Comment (felixbarny) [inline apm-agent-core/src/test/resources/json-specs/service_resource_inference.json:1]: 😍

**Comment by eyalkoren** [comment]
> The difference is that users don't need to remember setting resource fields or worry about which value to set in order for the span to show up in the service map.

Users will have to remember this new dedicated API and not the "default" one, and I think it's actually better to force them to think what the resource would be, instead of implicitly use `type` or `subtype`.

**Comment by eyalkoren** [comment]
As decided offline with @felixbarny - the exit span creation API is added.

**Pull request merged**